Adds metrics and log fields for tracking TLS handshake bytes

zwoop opened the pull request.

zwoop committed:
Capture TLS handshake bytes in metrics
Add txn log fields for the same HS metrics

Requestor: Craig T.

maskit commented on include/iocore/net/NetVConnection.h:
```diff
@@ -322,6 +322,24 @@ class NetVConnection : public VConnection, public PluginUserArgs<TS_USER_ARGS_VC
+   * @return true if values were captured on this call, false if already captured or not applicable
+   */
+  virtual bool
+  capture_tls_handshake_bytes(uint64_t &bytes_in, uint64_t &bytes_out)
```
I don't like to see TLS stuff in NetVConnection. `TLSBasicSupport` is probably a better place.

zwoop replied:
Hmmm, I changed the name of this already, it should have been capture_tls_handshake_bytes()

maskit commented on src/proxy/logging/LogAccess.cc:
```diff
@@ -2096,6 +2096,27 @@ LogAccess::marshal_client_req_squid_len(char *buf)
+    int64_t val = 0;
+
+    if (m_client_request) {
+      val = m_client_request->length_get() + m_http_sm->client_request_body_bytes;
```
What if the handshake has early data? I guess the size of the request gets double counted.

zwoop committed:
Move things out of the NetVC, subtract early data

maskit commented on doc/admin-guide/logging/formatting.en.rst:
```diff
@@ -451,6 +453,12 @@ cqcl  Client Request         Client request content length, in bytes.
 cqhl  Client Request         Client request header length, in bytes.
 cqql  Client Request         Client request header and content length combined,
                              in bytes.
+cqqtl Client Request         Same as cqql_, but for the first transaction on a
```
We may want a note (maybe as a limitation) about the accuracy. In some cases, the logged value wouldn't really reflect the amount of data transferred on wire to make the first transaction.

bneradt commented on doc/admin-guide/logging/formatting.en.rst:
```diff
@@ -451,6 +453,10 @@ cqcl  Client Request         Client request content length, in bytes.
                              in bytes.
+cqqtl Client Request         Same as cqql_, but for the first transaction on a
+                             TLS connection, also includes TLS handshake bytes
+                             received from the client. Note that this metrics
```
`this metrics` -> `this metric`

bneradt commented on src/iocore/net/SSLStats.cc:
```diff
@@ -280,6 +280,8 @@ SSLInitializeStatistics()
   ssl_rsb.user_agent_version_too_low        = Metrics::Counter::createPtr("proxy.process.ssl.user_agent_version_too_low");
   ssl_rsb.user_agent_wrong_version          = Metrics::Counter::createPtr("proxy.process.ssl.user_agent_wrong_version");
+  ssl_rsb.tls_handshake_bytes_in_total      = Metrics::Counter::createPtr("proxy.process.ssl.total_handshake_bytes_read_in");
+  ssl_rsb.tls_handshake_bytes_out_total     = Metrics::Counter::createPtr("proxy.process.ssl.total_handshake_bytes_write_in");
```
Seems to be a mismatch in variable name and the metric used. Should the variable be `in_total` or the metric pulled from `write_out`?

zwoop committed:
Fixed review comments

zwoop commented:
[approve ci autest 0]

zwoop committed:
Address Copilot review comments

- Use _tls_handshake_bytes_measured flag instead of > 0 sentinel,
  so the cache works correctly when both byte counts are zero
- Use std::min() when subtracting early data from handshake_rx to
  handle the == case and prevent underflow
- Add static_cast<int64_t> to marshal_int calls for rx/tx bytes

zwoop commented:
[approve ci centos]

cmcfarlen commented:
Cherry-picked to 10.2.x
